grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcHealthCheckService.java

```diff
@@ -0,0 +1,293 @@
+                                          .build();
+        this.grpcServiceHealthCheckers = grpcServiceHealthCheckers;
+        setInternalHealthUpdateListener(this.healthCheckers);
+        setInternalHealthUpdateListener(this.grpcServiceHealthCheckers.values());
+    }
+
+    /**
+     * Returns a newly created {@link GrpcHealthCheckService} with the specified {@link ListenableHealthChecker}s.
+     */
+    public static GrpcHealthCheckService of(ListenableHealthChecker... healthCheckers) {
+        return builder().checkers(healthCheckers).build();
+    }
+
+    /**
+     * Returns a newly created {@link GrpcHealthCheckService} with the specified {@link ListenableHealthChecker}s.
+     */
+    public static GrpcHealthCheckService of(Iterable<? extends ListenableHealthChecker> healthCheckers) {
+        return builder().checkers(healthCheckers).build();
+    }
+
+    /**
+     * Returns a new builder which builds a new {@link GrpcHealthCheckService}.
+     */
+    public static GrpcHealthCheckServiceBuilder builder() {
+        return new GrpcHealthCheckServiceBuilder();
+    }
+
+    @Override
+    public void check(HealthCheckRequest request, StreamObserver<HealthCheckResponse> responseObserver) {
+        String service = request.getService();
+        final Builder builder = HealthCheckResponse.newBuilder();
+        try {
+            builder.setStatus(checkServingStatus(service));
+        } catch (StatusException ex) {
+            responseObserver.onError(ex);
+            return;
+        }
+        responseObserver.onNext(builder.build());
+        responseObserver.onCompleted();
+    }
+
+    @Override
+    public void watch(HealthCheckRequest request,
+                      StreamObserver<HealthCheckResponse> responseObserver) {
+        String service = request.getService();
+        final Builder builder = HealthCheckResponse.newBuilder();
+        try {
+            builder.setStatus(checkServingStatus(service));
+        } catch (StatusException ex) {
+            builder.setStatus(ServingStatus.SERVICE_UNKNOWN);
+        }
+        HealthCheckResponse response = builder.build();
+        responseObserver.onNext(response);
+        if (watchers.containsKey(service)) {
+            watchers.get(service).put(responseObserver, response.getStatus());
+        } else {
+            final IdentityHashMap<StreamObserver<HealthCheckResponse>, ServingStatus> mp =
+                    new IdentityHashMap<>();
+            mp.put(responseObserver, response.getStatus());
+            watchers.put(service, mp);
+        }
+        ((ServerCallStreamObserver<HealthCheckResponse>) responseObserver).setOnCancelHandler(() -> {
+            if (watchers.containsKey(request.getService())) {
+                watchers.get(service).remove(responseObserver);
+                if (watchers.get(service).isEmpty()) {
+                    watchers.remove(service);
+                }
+            }
+        });
```

Comment:
I don't see we need to use `service` to find `responseObserver`.
Could we use `NonBlockingIdentityHashMap` instead of the nested hash map?

Reply:
If we need a service name in `provideInternalHealthUpdateListener()`, how about `NonBlockingIdentityHashMap<StreamObserver<HealthCheckResponse>, Entry<String, ServingStatus>>`?

Reply:
OK, I'll review and revisit the implementation.

Reply:
We need to call `checkServingStatus` in `provideInternalHealthUpdateListener`, so a service name as a key is necessary.
I'll use `NonBlockingIdentityHashMap<StreamObserver<HealthCheckResponse>, Entry<String, ServingStatus>>`